### PR TITLE
Prediction using linker

### PR DESCRIPTION
**Pull request by ibeltagy**
Adding a linker; a supervised sklearn classifier. 

With gold mentions and types:
```
MedMentions entities not in UMLS: 756
MedMentions entities found in UMLS: 91
K: 80, Filtered threshold : 1.0
Gold concept in candidates: 83.67%
Gold concept not in candidates: 16.28%
Doc level gold concept in candidates: 92.79%
Doc level gold concepts missed: 7.21%
Candidate generation failed: 0.05%
Correct mention-level linking: 63.52%  <<<===============
Mean, std, min, max candidate ids: 56.07%, 13.51%, 0, 164
Mean, std, min, max filtered candidate ids: 56.07%, 13.51%, 0, 164
```

With predicted mentions and types:
```
MedMentions entities not in UMLS: 756
MedMentions entities found in UMLS: 91
K: 80, Filtered threshold : 1.0
Gold concept in candidates: 51.90%
Gold concept not in candidates: 9.32%
Doc level gold concept in candidates: 75.51%
Doc level gold concepts missed: 24.49%
Candidate generation failed: 38.79%
Correct mention-level linking: 38.67%   <<<===============
Mean, std, min, max candidate ids: 55.97%, 13.63%, 0, 163
Mean, std, min, max filtered candidate ids: 55.97%, 13.63%, 0, 163
```

Will follow up with the code for generating training data and training classifier.

### CODE REVIEW

File: scripts/linking.py [hunk 1]
```diff
@@ -245,6 +245,7 @@ def load_tfidf_ann_index(model_path: str):
     # It makes the query slow without significant gain in recall.
     efS = 1000
     tfidf_vectorizer_path = f'{model_path}/tfidf_vectorizer.joblib'
+    linking_classifier_path = f'{model_path}/linking_classifier.joblib'
```
Comment (DeNeutoy) [inline scripts/linking.py]: This is conceptually separate from the candidate generation step, as we might train different classifiers. Could you separate the loading and saving of it from this function and make it optional in the top level script?

File: scripts/linking.py [hunk 2]
```diff
@@ -392,39 +421,62 @@ def eval_candidate_generation(examples: List[data_util.MedMentionExample],
...
-                    if len(filtered_ids) == 0:
+                    # Evaluating candidate generation
+                    if len(candidates) == 0:
                         entity_no_links_count += 1
                     elif gold_entity.umls_id in candidates:
                         entity_correct_links_count += 1
                     else:
                         entity_wrong_links_count += 1
 
+                    # Evaluating linking
```
Comment (DeNeutoy) [inline scripts/linking.py]: This should be it's own function outside the body of this loop, which does the classification for a single (mention, candidate_list) pair.

File: scripts/linking.py [hunk 3]
```diff
@@ -392,39 +421,62 @@ def eval_candidate_generation(examples: List[data_util.MedMentionExample],
...
+                    for candidate_id in candidate_ids:
+                        has_definition = 'definition' in umls_concept_dict_by_id[candidate_id]
+                        cosine_scores = [cosine for alias, cosine in candidates[candidate_id]]
+                        classifier_example = ({'has_definition': has_definition, 'cosines': cosine_scores,
+                                               'mention_types': mention_types,
+                                               'candidate_types': umls_concept_dict_by_id[candidate_id]['types'],})
+                        features.append(featurizer(classifier_example))
+                    if len(features) > 0:
+                        scores = linking_classifier.predict(features)
+                        pred_id = candidate_ids[np.argmax(scores)]
```
Comment (DeNeutoy) [inline scripts/linking.py]: Please can we compute recall@k here for the sorted list, for k in [1,3,5,10]

File: scripts/linking.py [hunk 4]
```diff
@@ -482,7 +535,7 @@ def main(medmentions_path: str,
         thresholds = [float(x) for x in thresholds.split(",")]
 
     # only evaluate on the dev examples for now because we don't have a trained model
-    eval_candidate_generation(dev_examples, umls_concept_dict_by_id, candidate_generator, k_list, thresholds, use_gold_mentions, spacy_model)
+    eval_candidate_generation_and_linking(dev_examples, umls_concept_dict_by_id, candidate_generator, linking_classifier, k_list, thresholds, use_gold_mentions, spacy_model)
```
Comment (DeNeutoy) [inline scripts/linking.py]: As above, can you make the linking classifier separate and optional.

File: scripts/linking.py (hunk 1 above)
Comment (ibeltagy) [reply scripts/linking.py]: done here https://github.com/allenai/scispacy/pull/106

File: scripts/linking.py (hunk 2 above)
Comment (ibeltagy) [reply scripts/linking.py]: added a Linker class (https://github.com/allenai/scispacy/pull/106)

File: scripts/linking.py (hunk 3 above)
Comment (ibeltagy) [reply scripts/linking.py]: will do in a separate PR

File: scripts/linking.py (hunk 4 above)
Comment (ibeltagy) [reply scripts/linking.py]: done

Comment (ibeltagy) [reply scripts/linking.py]: I made it optional and moved almost all its code in its own class, but it is still being called from this function. Moving it outside will result into a lot of duplicate code.

File: scripts/linking.py (hunk 3 above)
Comment (ibeltagy) [reply scripts/linking.py]: done here https://github.com/allenai/scispacy/pull/106

**Comment by ibeltagy** [comment]
> is this replaced by the other PR?
No, in addition to it.

**Pull request merged**